python-package/lightgbm/basic.py

```diff
@@ -594,6 +600,61 @@ def _load_pandas_categorical(file_name=None, model_str=None):
         return None
 
 
+class Sequence(object):
+    """Generic data access interface.
+
+    Object should support the following operations:
+
+        # Get total row number.
+        >>> len(seq)
+        # Random access by row index. Use for data sampling.
+        >>> seq[10]
+        # Range data access. Use to read data in batch when constructing Dataset.
+        >>> seq[0:100]
```

Comment:
I'm also considering use a separate interface such as `next_batch` to iterate over data in batch. So we don't need range access and there's no need to set batch size.

Reply:
comparison between current implementation sequence and `next`:

### Current
- LightGBM determines how to read, in-order out-of-order etc
- Data Provides hints reading behavior via `batch_size`
##### Potential improvements
- generally prefetching, ultimate form be like [auto-data-tunning ](https://www.tensorflow.org/guide/data_performance) in TensorFlow 
- parallel reading, even distributed reading
- natural to go with MPI (MPI I/O) Interface 

### `next`
- simplicity 
- data provider get to control what and how to read
- data provider could optimize data feeding accordingly (no need to wait for two segment of data from different cassette be fully loaded, just early cut when one is loaded and hint an prefetch event) 
##### future
- total separation of sampling and data reading stages, each could benefits in someway
- infinite dataset support 


To compare:
- both changes to current implementation and `next` approach could improve efficiency upon current implemntation
- current being more general, could leverage speed for almost any data source once sophisticated prefetch mechanism are implemented
- `next` being simplistic, giving all efficiency rights to data provider (note, data providers are still not limited in current implementation)

Reply:
just to note, these two implement do not conflict each other, in fact, to their ultimate goal, they could well co-exist to compensate each other.

Reply:
Is current approach more convenient and general for many potential accelerations? For prefetching, I think both current approach and the batch approach can do. But for parallel reading, it seems current approach is much more convenient, unless the data provider implements the `next` method with little overhead when a batch is not actually accessed. The current approach seems to put more burden on the data provider implementation though.

For prefetching when reading full dataset, I think we can pipeline reading from the `seq` and `push_rows` here.
https://github.com/cyfdecyf/LightGBM/blob/1c2a6ef185e813b11a1601764de09d5d576503cc/python-package/lightgbm/basic.py#L1575-L1580
If reading from the `seq` is I/O bounded, pipelining it with `push_rows` may hide the I/O latency.

Reply:
> Is current approach more convenient and general for many potential accelerations?
While I do think current approach is more general, it limits how data provider could optimize their own pipeline, since there is no order guarantee for them. On the other hand, LightGBM, since have the full control, can be optimized, like pipelining as you mentioned.

Idea here is, 
- If we trust user could build the best data provider against their own data storage traits, we should put limitations on the interface, so that data provider could get an garenteed data access pattern. This limits what lightGBM could do, yet maximizes global potential.
- If we don't, maybe we should keep using current general implementation, that we could do more (pipelining, prefetching, multiprocessing, distributed I/O, auto-tunnning, etc etc) and benefits everyone that implements it.
- I would say those two both has their benefits, if the user base is large enough, I would recommend implementing both. - (or even consider separate data sampling stage and parsing stage, allow the other one to be implemented in a total different way)

> For prefetching when reading full dataset, I think we can pipeline reading from the seq and push_rows here.

That seems like a very good idea. Are you implying multithreading? 

Large data reading could bounded by CPU (as long as they are not implemented in hardware offloaded fashion e.g. MPI), I'm not so sure if that would be the best.